specification/snapshot/_types/SnapshotRepository.ts

```diff
@@ -17,22 +17,119 @@
  * under the License.
  */
 
-import { Uuid } from '@_types/common'
+import { ByteSize, Uuid } from '@_types/common'
 import { integer } from '@_types/Numeric'
+import { Base } from '@xpack/usage/types'
+import { AdditionalProperties } from '@spec_utils/behaviors'
+import { UserDefinedValue } from '@spec_utils/UserDefinedValue'
+import { Stringified } from '@spec_utils/Stringified'
 
-export class Repository {
-  type: string
+export class BaseRepository<Settings> {
   uuid?: Uuid
-  settings: RepositorySettings
+  settings: Settings
 }
 
-export class RepositorySettings {
-  chunk_size?: string
-  compress?: string | boolean
-  concurrent_streams?: string | integer
+/**
+ * @variants internal tag='type'
+ * @non_exhaustive
+ */
+export type Repository =
+  | S3Repository
+  | GoogleCloudRepository
+  | AzureRepository
+  | SourceRepository
+  | UrlRepository
+  | FsRepository
+  | HdfsRepository
+
+export class S3Repository extends BaseRepository<S3RepositorySettings> {
+  type: 's3'
+}
+
+class BlobStoreSettings {
+  chunk_size?: ByteSize
+  concurrent_streams?: Stringified<integer>
+  compress?: Stringified<boolean>
+  max_number_of_snapshots?: integer
+  max_snapshot_bytes_per_sec?: ByteSize
+  max_restore_bytes_per_sec?: ByteSize
+  /** @aliases read_only */
+  readonly?: Stringified<boolean>
+  use_for_peer_recovery?: boolean
+}
+
+export class S3RepositorySettings {
+  base_path?: string
+  bucket?: string
+  buffer_size?: ByteSize
+  client?: string
   location: string
-  /**
-   * @aliases readonly
-   */
-  read_only?: string | boolean
+  server_size_encryption?: boolean
+  storage_class?: string
+}
+
+export class AzureRepository extends BaseRepository<AzureRepositorySettings> {
+  type: 'azure'
+}
+
+export class AzureRepositorySettings extends BlobStoreSettings {
+  /** aliases account */
+  client?: string
+  container?: string
+  location_mode: AzureRepositoryLocation
+  max_single_part_upload_size: ByteSize
+}
+
+export enum AzureRepositoryLocation {
+  primary_only,
+  secondary_only,
+  primary_then_secondary,
+  secondary_then_primary
+}
+
+export class GoogleCloudRepository extends BaseRepository<GoogleCloudRepositorySettings> {
+  type: 'gcs'
+}
+
+export class GoogleCloudRepositorySettings extends BlobStoreSettings {
+  bucket?: string
+  base_path?: string
+}
+
+export class UrlRepository extends BaseRepository<UrlRepositorySettings> {
+  type: 'url'
+}
+
+export class UrlRepositorySettings {
+  url: string
+  supported_protocols?: string
+  allowed_urls?: string[]
+}
+
+export class SourceRepository extends BaseRepository<SourceRepositorySettings> {
+  type: 'source'
+}
+
+export class SourceRepositorySettings
+  implements AdditionalProperties<string, UserDefinedValue>
+{
+  delegate_type: string
+}
+
+export class FsRepository extends BaseRepository<FsRepositorySettings> {
+  type: 'fs'
+}
+
+export class FsRepositorySettings extends BlobStoreSettings {
```

Comment:
From docs, this is missing:

- chunk_size
- compress
- max_number_of_snapshots
- max_restore_bytes_per_sec
- max_snapshot_bytes_per_sec
- readonly

Reply:
They are inherited from `BlobStoreSettings`